### ISSUE TITLE
DST adjustments to schedules miss behave when day pattern changes right at midnight. (CR #8026)

### ISSUE THREAD

**Issue by axelstudios** [NREL/EnergyPlus#2804]
## 
#### Description

[3/9/2010 BG]  user has work around.  may contribute NFP/EL item on this.  issue was related to different understandings of day pattern concept versus literal time series. verifying as designed

disgruntled/annoying user has compliant about DST adjustment to schedules when schedules vary during the 24:00 to 1:00 am timeframe.  May have a point so I am filing this CR. 

Defect files uploaded.   8026*

Copy of Email thread. 
_______________________________________--
All,

One more simple (and contrived) example to illustrate my point. Source files and output files are attached and relevant excerpts copied below. Recall that all reported times are in standard time, so you will have to mentally shift everything ahead by one hour to understand when it was simulated in daylight time (only with DST is enabled!).
You should see that HVAC operation (HVACOperationSchd), scheduled for hour 1:00 on June 2 is reported at hour 24:00 on June 2. When you shift the output to daylight saving time, you will see that the HVAC system operates at hour 1:00 of June 3. Absolutely, unequivocally, wrong!

Chad
## HVAC Schedule

Schedule:Compact,
  HVACOperationSchd,       !- Name
  On/Off,                  !- Schedule Type Limits Name
  Through: 6/1,            !- Field 1
  For: AllDays,            !- Field 2
  Until: 24:00,            !- Field 3
  0.0,                     !- Field 4
  Through: 6/2,            !- Field 5
  For: AllDays,            !- Field 6
  Until: 1:00,             !- Field 7
  1.0,                     !- Field 8
  Until: 24:00,            !- Field 9
  0.0,                     !- Field 10
  Through: 12/31,          !- Field 11
  For: AllDays,            !- Field 12
  Until: 24:00,            !- Field 13
  0.0;                     !- Field 14
## Output from simulation with DST enabled

Date/Time,Electricity:HVAC [J](Hourly)
 06/01  01:00:00,0.0
 06/01  02:00:00,0.0
 ...
 06/01  23:00:00,0.0
 06/01  24:00:00,0.0
 06/02  01:00:00,0.0
 06/02  02:00:00,0.0
 ...
 06/02  23:00:00,0.0
 06/02  24:00:00,7111360.40038745
 06/03  01:00:00,0.0
 06/03  02:00:00,0.0
 ...
 06/03  23:00:00,0.0
 06/03  24:00:00,0.0

Best,
Chad

On Mon, Feb 22, 2010 at 7:27 PM, Chad Corbin chad.corbin@gmail.com wrote:

> Hi Kristin,
> 
> Thanks for the response.
> 
> Yep, we know everything is reported in Standard Time; that isn't the 
> issue at all. The issue is the order of scheduled values. This 
> manifests when we want a different schedule for each hour of each day.
> Believe me, we've done plenty of testing and spent more than enough 
> time on this! Nothing blows-up, but you do see the first hour of 
> Sunday's schedule simulated at the end of Sunday, immediately before 
> Monday's schedule. This simply doesn't make sense.
> 
> We are dynamically writing schedules to the IDF from another program 
> to test control strategies, but identical control strategies will be 
> simulated differently by E+ depending on the time of year. In order 
> for us to get around this we have to do one of two things:
> 
> 1) eliminate daylight saving from the simulation and manually shift 
> the weather file by one hour during daylight saving (although this 
> doesn't address solar angle calculations).
> 
> 2) write the schedule value we want simulated in the first hour of 
> Monday's schedule in the first hour of Sunday's schedule instead.
> 
> Neither is perfect. It would be better if E+ simulated the schedule as 
> written instead of breaking it up and rearranging it as illustrated in 
> the short example below. The building goes from being operated for 
> eight hours continuously (standard time) to being operated for seven 
> hours continuously, then again for one hour at the end of the day 
> (daylight time). Most users won't notice this because schedule values 
> at 11pm are often the same as those at midnight and 1am. They aren't 
> in our case.
> 
> BTW, both examples are taken directly from a set of E+ input and output files.
> 
> Thanks!
> Chad
> 
> On Mon, Feb 22, 2010 at 5:14 PM, Field, Kristin Kristin.Field@nrel.gov wrote:
> 
> > Hey, Chad.  I was wrong before - it's not a bug.  It's just not the most intuitive way of dealing with daylight savings.  Here is what is actually happening (from I/O ref):
> > 
> > "Note: For EnergyPlus Output:Variable and Output:Meter reporting, the 
> > time stamps are always in standard time. When daylight saving time is 
> > active, scheduled loads and controls will shift one hour relative to standard time."
> > 
> > Of course, it would make more sense to skip 24.101 and use 24.201 at hour 24 of day 1.  Then use 24.301 at hour 24 of day 2, etc.  Since your schedule is the same for all day types (i.e., "AllDays"), then I can see why this wouldn't really matter.  I wonder if it would blow up if you had different schedules for weekends and weekdays, though, or especially a different one for each day of the week.  If you have some intellectual curiosity on the matter, try it and let us know!  Otherwise, I think this at least won't affect your results since all the schedules are being modified the same way.
> > 
> > -----Original Message-----
> > From: Chad Corbin [mailto:chad.corbin@gmail.com]
> > Sent: Sunday, February 21, 2010 5:41 PM
> > To: Margarete Rois Langner; Neal Kruis; Field, Kristin; Horowitz, 
> > Scott
> > Subject: Issue regarding E+ schedules and daylight saving
> > 
> > Howdy folks!
> > 
> > Since you are experts in all things E+, I thought I'd run something 
> > by you regarding E+ schedules during daylight saving. I'm interested 
> > in your opinion, and/or that of your expert co-workers not included here.
> > 
> > The short description:
> > 
> > All schedule values defined between midnight and 1:00AM will be 
> > simulated 23 hours later between 11:00PM and midnight when daylight 
> > saving is enabled.
> > 
> > The short example:
> > 
> > An HVAC schedule is defined as 1.0 between midnight and 8:00AM on 
> > Sundays only. All other days and hours have a schedule value of 0.0.
> > 
> > On February 1, the HVAC equipment will be enabled at midnight (local
> > time) on Sunday and run for 8 hours. On June 1, the HVAC equipment 
> > will be enabled at midnight (local time) on Sunday, run for 7 hours, 
> > shut-off, then run again between 11:00PM and midnight.
> > 
> > The full blown description and example:
> > 
> > When E+ simulates a day during daylight saving, it takes the schedule 
> > value for the first hour of the day, ie. 'Until 1:00', removes it, 
> > shifts the other hours back by one (as expected), and implements the 
> > first hour value at the 24th hour of the same day (not expected). The 
> > behavior appears for all days in the simulation period. Therefor, any 
> > schedule having a unique value between midnight and 1:00AM will have 
> > that value implemented 23 hours later at 11:00PM on the same day.
> > 
> > This can be seen in the example schedule and E+ output included below.
> > In this example, two days were scheduled where each hour had a unique 
> > setpoint value. Setpoints were constructed so that the values 
> > represented the hour and day of simulation. The tenths decimal 
> > represents the day (1 or 2) and the hundredths and thousandths 
> > decimal represent the hour (01 through 24). For example, 10pm on the 
> > second day has a setpoint of 24.220.
> > 
> > The schedule was simulated with daylight savings enabled via the 
> > RunPeriodControl:DaylightSavingTime object, and the results shown 
> > below. All times are shown in standard time. Note that the last hour 
> > of the day shows the first setpoint value of the same day:
> > 
> > 06/01  24:00:00 24.101
> > 
> > The same can also be seen in the last hour of the second day:
> > 
> > 06/02  24:00:00 24.201
> > 
> > Over the entire daylight savings period, one would expect to see the 
> > first hour in the schedule of the first day thrown away, the first 
> > hour of the next day implemented at the last hour of the first day, 
> > and the last hour of the last day repeated. This does not appear to 
> > be the case. Instead, this occurs every day of the simulation.
> > 
> > For any period wholly contained within the daylight savings period, I 
> > would expect to see similar behavior whereby the first hour in the 
> > schedule of day X+1 is implemented in the last hour of day X. The 
> > last hour of the last day in the simulation period should be the next 
> > hour defined in the schedule.
> > 
> > The schedule definition:
> > 
> > Schedule:Compact,
> > CLGSETP_SCH, Temperature,
> > Through: 6/1, For: AllDays,
> > Until: 1:00,
> > 24.101,
> > Until: 2:00,
> > 24.102,
> > Until: 3:00,
> > 24.103,
> > Until: 4:00,
> > 24.104,
> > Until: 5:00,
> > 24.105,
> > .
> > .
> > .
> > Until: 21:00,
> > 24.121,
> > Until: 22:00,
> > 24.122,
> > Until: 23:00,
> > 24.123,
> > Until: 24:00,
> > 24.124,
> > Through: 6/2, For: AllDays,
> > Until: 1:00,
> > 24.201,
> > Until: 2:00,
> > 24.202,
> > Until: 3:00,
> > 24.203,
> > Until: 4:00,
> > 24.204,
> > Until: 5:00,
> > 24.205,
> > .
> > .
> > .
> > Until: 21:00,
> > 24.221,
> > Until: 22:00,
> > 24.222,
> > Until: 23:00,
> > 24.223,
> > Until: 24:00,
> > 24.224;
> > 
> > The E+ CSV Output:
> > 
> > 06/01  01:00:00 24.102
> > 06/01  02:00:00 24.103
> > 06/01  03:00:00 24.104
> > 06/01  04:00:00 24.105
> > .
> > .
> > .
> > 06/01  21:00:00 24.122
> > 06/01  22:00:00 24.123
> > 06/01  23:00:00 24.124
> > 06/01  24:00:00 24.101 (?)
> > 06/02  01:00:00 24.202
> > 06/02  02:00:00 24.203
> > 06/02  03:00:00 24.204
> > 06/02  04:00:00 24.205
> > .
> > .
> > .
> > 06/02  21:00:00 24.222
> > 06/02  22:00:00 24.223
> > 06/02  23:00:00 24.224
> > 06/02  24:00:00 24.201 (?)
> > 
> > Does this seem right to you?
> > 
> > Thanks,
> > Chad

LKL, 3/8/2010
Reporting the HVAC Schedule Value illustrates that the operation of the HVAC is scheduled for 0:10 to 1:00 on non-DST (6/2) and from 23:10 to 24:00 on DST (6/2) which is as it seems it should be.  Not sure where he's getting the idea for 6/3 in this.  Unless there is some other lag going on that would make the HVAC system operate then?  Temperatures obviously could vary.

Sent him this message:
Chad,

I have spent considerable time looking at this as well as creating an input file of my own to illustrate.
Mine is a very simple schedule that should show the day type number for each day.

I changed your reporting from hourly to "detailed" to futher illustrate as you have to be cautious in using hourly when you are reporting an averaged (over the time step) variable.

I reported out, for your inputs -- the ones you sent me and the ones you sent Brent -- the DST indicator, the day type number, and the schedule value.  Plus some of yours also report Electricity on the hourly value.  My file does the same but reports the schedule value which should be the day type number. (EnergyPlus convention: Sunday = 1)

Your schedule excerpt:
  Through: 6/2,            !- Field 5
  For: AllDays,            !- Field 6
  Until: 1:00,             !- Field 7
  1.0,                     !- Field 8
  Until: 24:00,            !- Field 9
  0.0,                     !- Field 10

on 6/2 for the DST versions of your file, I see the 1 appearing at hours 23:10 through 24:00, as expected.
for the non-DST versions, I see the 1 appearing at 0:10 through 1:00.

I believe that is what I said the expected behavior was for EnergyPlus.  I do not see it appearing in the following day (6/3).

In other words, I do not believe there is a problem in EnergyPlus.

Attached are the five input files and the five csv outputs -- in a zip file with the extension renamed to ~~~ to protect certain people who are on the cc list.

Linda

and adding the zip file to the defectfiles.
## 

Addressed by Linda Lawrie
Last build tested: `10.02.24 V5.0.0.019`
